tiledb/sm/array_schema/array_schema.cc

```diff
@@ -763,37 +764,50 @@ Status ArraySchema::add_dimension_label(
     if (name != dim->name()) {
       // Check no attribute with this name
       bool has_matching_name{false};
-      RETURN_NOT_OK(has_attribute(name, &has_matching_name));
-      if (has_matching_name)
-        return LOG_STATUS(Status_ArraySchemaError(
+      throw_if_not_ok(has_attribute(name, &has_matching_name));
+      if (has_matching_name) {
+        throw ArraySchemaStatusException(
             "Cannot add a dimension label with name '" + std::string(name) +
-            "'. An attribute with that name already exists."));
+            "'. An attribute with that name already exists.");
+      }
+
       // Check no dimension with this name
-      RETURN_NOT_OK(domain_->has_dimension(name, &has_matching_name));
-      if (has_matching_name)
-        return LOG_STATUS(Status_ArraySchemaError(
+      throw_if_not_ok(domain_->has_dimension(name, &has_matching_name));
+      if (has_matching_name) {
+        throw ArraySchemaStatusException(
             "Cannot add a dimension label with name '" + std::string(name) +
-            "'. A different dimension with that name already exists."));
+            "'. A different dimension with that name already exists.");
+      }
     }
+
     // Check no other dimension label with this name.
     auto found = dimension_label_map_.find(name);
-    if (found != dimension_label_map_.end())
-      return LOG_STATUS(Status_ArraySchemaError(
+    if (found != dimension_label_map_.end()) {
+      throw ArraySchemaStatusException(
           "Cannot add a dimension label with name '" + std::string(name) +
-          "'. A different label with that name already exists."));
+          "'. A different label with that name already exists.");
+    }
   }
-  // Check the datatype of the dimension label is consistent with the dimension
-  // it is being added to.
-  if (check_is_compatible && !dimension_label_schema->is_compatible_label(dim))
-    return LOG_STATUS(Status_ArraySchemaError(
-        "Cannot add dimension label; The dimension label schema is not "
-        "compatible with the dimension it is being added to."));
-  // Create relative URI in dimension label directory
-  URI uri{constants::array_dimension_labels_dir_name + "/l" +
-              std::to_string(nlabel_internal_),
-          false};
+
   // Add dimension label
   try {
+    // Create dimension label schema.
+    // - Note: This is a placeholder until dimension label schema is replace
+    // directly with an array schema.
+    auto dimension_label_schema = make_shared<DimensionLabelSchema>(
+        HERE(),
+        label_order,
+        label_type,
+        dim->type(),
+        dim->domain().data(),
+        dim->tile_extent().data());
+
+    // Create relative URI in dimension label directory
+    URI uri{constants::array_dimension_labels_dir_name + "/l" +
```

Comment:
Note: needs to go into `storage_format` when this code is moved out of experimental.